clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -4298,6 +4296,28 @@ void handleFailure(Throwable throwable) {
                         }
                     }
                 }
+
+                @Override
+                boolean handleUnsupportedVersionException(UnsupportedVersionException exception) {
+                    if (supportsMaxTimestamp) {
+                        supportsMaxTimestamp = false;
+
+                        // fail any unsupported futures
+                        partitionsToQuery.stream().forEach(
+                            t -> t.partitions().stream()
+                                .filter(p -> p.timestamp() == ListOffsetsRequest.MAX_TIMESTAMP)
+                                .forEach(
+                                    p -> futures.get(new TopicPartition(t.name(), p.partitionIndex()))
+                                        .completeExceptionally(
+                                            new UnsupportedVersionException(
+                                                "Broker " + brokerId
+                                                    + " does not support MAX_TIMESTAMP offset spec"))
+                                )
+                        );
```

Comment:
Should we removed the `MAX_TIMESTAMP` ones from `partitionsToQuery` as well? Otherwise, they will be retried.

Reply:
good call, I've updated to do this.